src/main/scala/com/microsoft/ml/spark/vw/VowpalWabbitContextualBandit.scala

```diff
@@ -0,0 +1,247 @@
+
+    schema
+  }
+
+  protected override def trainRow(schema: StructType,
+                                  inputRows: Iterator[Row],
+                                  ctx: TrainContext
+                                 ): Unit = {
+    val allActionFeatureColumns = Seq(getFeaturesCol) ++ getAdditionalFeatures
+    val allSharedFeatureColumns = Seq(getSharedCol) ++ getAdditionalSharedFeatures
+
+    val actionNamespaceInfos = VowpalWabbitUtil.generateNamespaceInfos(
+      schema,
+      getHashSeed,
+      allActionFeatureColumns)
+
+    val sharedNamespaceInfos = VowpalWabbitUtil.generateNamespaceInfos(schema, getHashSeed, allSharedFeatureColumns);
+    val chosenActionColIdx = schema.fieldIndex(getChosenActionCol)
+    val labelIdx = schema.fieldIndex(getLabelCol)
+    val probabilityIdx = schema.fieldIndex(getProbabilityCol)
+
+    val exampleStack = new ExampleStack(ctx.vw)
+
+    for (row <- inputRows) {
+      VowpalWabbitUtil.prepareMultilineExample(row, actionNamespaceInfos, sharedNamespaceInfos, ctx.vw, exampleStack,
+        examples => {
+          // It's one-based but we need to skip the shared example anyway
+          val selectedActionIdx = row.getInt(chosenActionColIdx)
+          if (selectedActionIdx == 0) {
+            throw new IllegalArgumentException("Chosen action index is 1 based - cannot be 0")
+          }
+
+          val cost = row.getDouble(labelIdx)
+          val loggedProbability = row.getDouble(probabilityIdx)
+
+          // Set the label for learning
+          examples(selectedActionIdx).setContextualBanditLabel(
+            selectedActionIdx,
+            cost,
+            loggedProbability)
+
+          // Learn from the examples
+          ctx.vw.learn(examples)
+
+          // Update the IPS/SNIPS estimator
+          val prediction: ActionProbs = examples(0).getPrediction.asInstanceOf[ActionProbs]
+          val probs = prediction.getActionProbs
+          val selectedActionIdxZeroBased = selectedActionIdx - 1
+          ctx.contextualBanditMetrics.add_example(loggedProbability, cost, probs.apply(selectedActionIdxZeroBased).getProbability)
+        })
+    }
+  }
+
+  override protected def train(dataset: Dataset[_]): VowpalWabbitContextualBanditModel = {
+    val model = new VowpalWabbitContextualBanditModel(uid)
+      .setFeaturesCol(getFeaturesCol)
+      .setAdditionalFeatures(getAdditionalFeatures)
+      .setSharedCol(getSharedCol)
+      .setPredictionCol(getPredictionCol)
+
+    trainInternal(dataset, model)
+  }
+
+  override def copy(extra: ParamMap): VowpalWabbitContextualBandit = defaultCopy(extra)
+}
+
+@InternalWrapper
+class VowpalWabbitContextualBanditModel(override val uid: String)
+  extends PredictionModel[Row, VowpalWabbitContextualBanditModel]
+    with VowpalWabbitBaseModel with VowpalWabbitContextualBanditBase {
+
+  override def transformSchema(schema: StructType): StructType = schema
+
+  lazy val exampleStack = new ExampleStack(vw)
```

Comment:
if i remember this was mutable, we have to make sure that workers arent adding to this (because that wont work on cluster) and we should probably discuss this a bit anyway